resources-manager/include/AdaptivePruningManager.hpp

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include "IRequestHandler.hpp"
+#include "IResourceManager.hpp"
+#include "InternalBFTClient.hpp"
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <thread>
+
+enum PruningMode { LEGACY, ADAPTIVE };
+
+namespace concord::performance {
+class AdaptivePruningManager {
+ public:
+  AdaptivePruningManager(const std::shared_ptr<concord::performance::IResourceManager> &resourceManager,
+                         const std::chrono::duration<double, std::milli> &interval,
+                         const std::shared_ptr<bftEngine::impl::IInternalBFTClient> &bftClient);
+  virtual ~AdaptivePruningManager();
+
+  void switchMode(PruningMode newMode) {
+    mode = newMode;
+    conditionVar.notify_one();
+  }
+  PruningMode getCurrentMode() { return mode.load(); }
+
+  void youArePrimary() {  // Is there a listener for becoming a primary replica?
```

Comment:
You have it in the replicaImp file. Basically, the primary is always the current_view % num_replicas.
I would suggest that replicaImp will call this method once the view is changed